### PR TITLE
JS: Add `dot.js` support

### PR DESCRIPTION
**Pull request by jorgectf**

### CODE REVIEW

**Comment by asgerf** [comment]
Looks reasonable, but I'm thinking there are probably a few other types of `.dot` files that aren't part of this templating system, such as Graphviz dot. We'll need to make sure that it has no adverse effects when the repository contains such files. I don't think I'll have the time to investigate this week, however, so apologies in advance for leaving this PR hanging for a bit, as another issue is currently taking priority.

File: javascript/ql/lib/semmle/javascript/frameworks/Templating.qll [hunk 1]
```diff
@@ -580,6 +580,22 @@ module Templating {
     override string getAPackageName() { result = "ejs" }
   }
 
+  /**
+   * doT-style syntax, using `{{! }}` for safe interpolation, and `{{= }}` for
+   * unsafe interpolation.
```
Comment (asgerf) [inline javascript/ql/lib/semmle/javascript/frameworks/Templating.qll:585]: Looks like the two escaping mechanisms were flipped around, either in the comment or in the implementation.

- The comment says `{{! }}` is safe, but the implementation treats it as raw (i.e. unsafe)
- The comment says `{{= }}` is unsafe, but the implementation treats it as escaping (i.e. safe, in most contexts).

**Review by asgerf — COMMENTED** [review]
A quick [code search](https://github.com/search?q=path%3A**%2F*.dot+&type=code) shows about 800k files with the `.dot` extension, and from [a few](https://github.com/search?q=path%3A**%2F*.dot+%22%3Cdiv%22&type=code) [other](https://github.com/search?q=path%3A**%2F*.dot+%22%3Cspan%22&type=code) [searches](https://github.com/search?q=path%3A**%2F*.dot+%22%3Chtml%22&type=code) I'd estimate around 1k of those are DoT templates.

Based on that, I'd like to restrict the extraction a bit to avoid extracting so many unrelated files as HTML.

For `.erb` files we only actually extract `.html.erb` files (due to [this code](https://github.com/github/codeql/pull/13624/files#diff-fb2c06c9c89aba8fdc094f64e28c6f4bc6379b063ea9be942a179221c31f8254R123)). Would a similar restriction work for your use case, i.e. restricting to just `.html.dot` files?

Otherwise we can sniff the file header for HTML tags but that's a bit more work.

File: javascript/ql/lib/semmle/javascript/frameworks/Templating.qll (hunk 1 above)
Comment (jorgectf) [reply javascript/ql/lib/semmle/javascript/frameworks/Templating.qll:585]: Oops, sorry, the model is right, `{{! }}` is for unsafe (raw) interpolation.

**Comment by jorgectf** [comment]
> Based on that, I'd like to restrict the extraction a bit to avoid extracting so many unrelated files as HTML.
> 
> For `.erb` files we only actually extract `.html.erb` files (due to [this code](https://github.com/github/codeql/pull/13624/files#diff-fb2c06c9c89aba8fdc094f64e28c6f4bc6379b063ea9be942a179221c31f8254R123)). Would a similar restriction work for your use case, i.e. restricting to just `.html.dot` files?
> 
> Otherwise we can sniff the file header for HTML tags but that's a bit more work.

Thanks for the analysis! I'm happy to restrict it to `.html.dot` to avoid extracting unrelated files. wdyt @Kwstubbs?

**Comment by asgerf** [comment]
@jorgectf any updates? Let me know if you need to help from my end. I'd prefer to get the PR merged rather than hanging.

**Comment by Kwstubbs** [comment]
@asgerf What do you think about adding the (.jst)[https://github.com/search?q=path%3A**%2F*.jst&type=code] extension as well? Seems this is used by other template engines as well.

**Comment by asgerf** [comment]
@Kwstubbs I'd prefer to focus on getting the PR merged, without expanding its scope. JST can be added in a separate PR.

**Comment by Kwstubbs** [comment]
@asgerf made the same changes as the .html.erb files, let me know if that will do the trick.

**Comment by asgerf** [comment]
Unfortunately the test doesn't seem to react to the changes. There should be an update to the corresponding `.expected` file showing the new data flow into the template, and currently it doesn't find that flow.

I've pushed two commits that fix two of the underlying issues. For some reason the test still isn't finding the flow, so there's still some work to do here, though I need to commit my time elsewhere right now.

**Comment by jorgectf** [comment]
Thank you @asgerf for your dedication, I'll work with @Kwstubbs to try to figure things out.

**Comment by asgerf** [comment]
The issue might be that `TEMPLATE_EXPR_OPENING_TAG` needs to include `{{!` as the opening of an expression tag.

**Comment by jorgectf** [comment]
> The issue might be that `TEMPLATE_EXPR_OPENING_TAG` needs to include `{{!` as the opening of an expression tag.

The tests still don't find the new results :/

**Comment by erik-krogh** [comment]
> The tests still don't find the new results :/

I found the reason. I've pinged you in an internal PR. 
There is also a backref to the internal PR.  

Your tests work after that PR has been merged.

**Comment by erik-krogh** [comment]
Hmm. 
The `javascript/ql/test/library-tests/frameworks/Templating/Xss.qlref` and `javascript/ql/test/library-tests/frameworks/Templating/test.ql` tests are failing in the CI.  
Can you look into that.  

I did some evaluations. They look fine and were very uneventful.

**Comment by jorgectf** [comment]
> Hmm. The `javascript/ql/test/library-tests/frameworks/Templating/Xss.qlref` and `javascript/ql/test/library-tests/frameworks/Templating/test.ql` tests are failing in the CI. Can you look into that?
> 
> I did some evaluations. They look fine and were very uneventful.

I haven't been able to figure out a fix for the error. I think it has to be with the internal test, which I'm not familiar with.

**Comment by erik-krogh** [comment]
> I haven't been able to figure out a fix for the error. I think it has to be with the internal test, which I'm not familiar with.

It's because you have to compile the extractor with your changes to see the changes to the tests.   
I've done that and pushed the updated test outputs here.  

I think those outputs are what we should expect from your changes?

**Comment by jorgectf** [comment]
> > I haven't been able to figure out a fix for the error. I think it has to be with the internal test, which I'm not familiar with.
> 
> It's because you have to compile the extractor with your changes to see the changes to the tests. I've done that and pushed the updated test outputs here.
> 
> I think those outputs are what we should expect from your changes?

Yes, the expected results LGTM, thank you!

**Review by erik-krogh — APPROVED** [review]
Great.  
LGTM 👍 

@asgerf: Do you have any comments?

**Comment by asgerf** [comment]
Perhaps a final evaluation, otherwise LGTM, and thanks for driving this to completion!

**Comment by erik-krogh** [comment]
> Perhaps a final evaluation, otherwise LGTM, and thanks for driving this to completion!

Final evaluations (see backrefs) were very uneventful. 
They were on nightly and default, and with building the extractor from `github/codeql`.  

I'm merging.

**Pull request merged**